### ISSUE TITLE
Use select.poll instead of select.select in SocketServer.BaseServer.serve_forever

### ISSUE THREAD

**Issue by 485ee0c2-dd9f-4240-b0ed-198b733495a1** [python/cpython#75791]
BPO | [31610](https://bugs.python.org/issue31610)
--- | :---
Nosy | @vstinner, @giampaolo

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2017-09-27.11:24:44.981>
labels = ['library', 'performance']
title = 'Use select.poll instead of select.select in SocketServer.BaseServer.serve_forever'
updated_at = <Date 2017-09-28.15:05:48.447>
user = 'https://bugs.python.org/bboneva'
```

bugs.python.org fields:
```python
activity = <Date 2017-09-28.15:05:48.447>
actor = 'vstinner'
assignee = 'none'
closed = False
closed_date = None
closer = None
components = ['Library (Lib)']
creation = <Date 2017-09-27.11:24:44.981>
creator = 'bboneva'
dependencies = []
files = []
hgrepos = []
issue_num = 31610
keywords = []
message_count = 4.0
messages = ['303127', '303247', '303251', '303254']
nosy_count = 3.0
nosy_names = ['vstinner', 'giampaolo.rodola', 'bboneva']
pr_nums = []
priority = 'normal'
resolution = None
stage = None
status = 'open'
superseder = None
type = 'performance'
url = 'https://bugs.python.org/issue31610'
versions = []
```

</p></details>

**Comment by 485ee0c2-dd9f-4240-b0ed-198b733495a1** [comment]
The select function does not work for file descriptors with number \>= 1024. It has FD_SETSIZE limit (see https://stackoverflow.com/questions/7976388/increasing-limit-of-fd-setsize-and-select#7977082).
So, serve_forever won't work if more than 1024 file descriptors are opened in the process (ValueError: filedescriptor out of range in select() is raised).
Moreover, the select function is considered inefficient, hard to scale and old-fashioned nowadays. Refer to https://stackoverflow.com/questions/970979/what-are-the-differences-between-poll-and-select#3951845.
Better alternative will be to use poll or epoll even.

**Comment by vstinner** [comment]
> Better alternative will be to use poll or epoll even.

epoll requires mulitple system calls, whereas the selector seems to only be used only once in most cases, no?

Anyway, maybe selectors.DefaultSelector can be used here?

About poll(): be aware of macOS issues with poll() :-/

**Comment by giampaolo** [comment]
I recommend to use selectors.PollSelector and fallback on selectors.SelectSelector where not available (Windows) in order to use 1 syscall only. That's what we did already in socket.py, subprocess.py and others.

**Comment by vstinner** [comment]
> I recommend to use selectors.PollSelector and fallback on selectors.SelectSelector where not available (Windows) in order to use 1 syscall only. That's what we did already in socket.py, subprocess.py and others.

It would be nice to have an API in selectors to get a "lightweight" selector, for selector and poll.

This API would be appropriate to check on macOS if poll() works or not: bpo-28087.
